internal/connector/destination.go

```diff
@@ -287,59 +233,213 @@ func (d *Destination) Write(ctx context.Context, dstCfgPath string, cfgCatalogPa
+			d.logger.Log(airbyte.LogLevelError, fmt.Sprintf("Append destination sync mode is not compatible with Data Pool %q ORDER BY statement", dataSource.ID))
+			return fmt.Errorf("append destination sync mode is not compatible with Data Pool %q ORDER BY statement", dataSource.ID)
+		}
+	}
+
+	if err := d.writeRecords(ctx, input, dataSources); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *Destination) buildAndCreateDataSource(ctx context.Context, configuredStream airbyte.ConfiguredStream, dataSourceUniqueName string, apiClient PropelApiClient) (*models.DataSource, error) {
+	// Generates a password of 18 chars length with 2 digits, 2 symbols and uppercase letters.
+	authPassword, err := password.Generate(18, 2, 2, false, false)
+	if err != nil {
+		d.logger.Log(airbyte.LogLevelError, fmt.Sprintf("Password generation failed: %v", err))
+		return nil, fmt.Errorf("failed to generate Basic auth password for Data Source %q: %w", dataSourceUniqueName, err)
+	}
+
+	orderByColumns := make([]string, 0, len(configuredStream.PrimaryKey))
+	for _, pk := range configuredStream.PrimaryKey {
+		if len(pk) != 1 {
+			d.logger.Log(airbyte.LogLevelError, fmt.Sprintf("Unexpected primary key length %d for Data Source %q", len(pk), dataSourceUniqueName))
+			return nil, fmt.Errorf("unexpected primary key length %d for Data Source %q", len(pk), dataSourceUniqueName)
+		}
+
+		orderByColumns = append(orderByColumns, pk[0])
+	}
+
+	var cursorField string
+	if len(configuredStream.CursorField) > 0 {
+		cursorField = configuredStream.CursorField[0]
+	}
+
+	baseColumns := make([]*models.WebhookDataSourceColumnInput, 0, len(configuredStream.Stream.JSONSchema.Properties)+len(defaultAirbyteColumns))
+
+	for propertyName, propertySpec := range configuredStream.Stream.JSONSchema.Properties {
+		columnType, err := ConvertAirbyteTypeToPropelType(propertySpec.PropertyType)
+		if err != nil {
+			d.logger.Log(airbyte.LogLevelError, fmt.Sprintf("Airbyte to Propel data type conversion failed for Data Source %q: %v", dataSourceUniqueName, err))
+			return nil, fmt.Errorf("failed to convert Airbyte to Propel data type: %w", err)
+		}
+
+		baseColumns = append(baseColumns, &models.WebhookDataSourceColumnInput{
+			Name:         propertyName,
+			Type:         columnType,
+			Nullable:     !slices.Contains(orderByColumns, propertyName) && propertyName != cursorField,
+			JsonProperty: propertyName,
+		})
+	}
+
+	createDataSourceOpts := client.CreateDataSourceOpts{
+		Name: dataSourceUniqueName,
+		BasicAuth: &models.HttpBasicAuthInput{
+			Username: configuredStream.Stream.Namespace,
+			Password: authPassword,
+		},
+		Columns: baseColumns,
+	}
+
+	if len(orderByColumns) == 0 && configuredStream.DestinationSyncMode == airbyte.DestinationSyncModeAppendDedup {
+		d.logger.Log(airbyte.LogLevelError, fmt.Sprintf("Append Dedup sync mode requires at least 1 primary key"))
```

Comment:
Nit: the primary key is the set of columns, so we need at least one primary key _column_, no?
```suggestion
		d.logger.Log(airbyte.LogLevelError, fmt.Sprintf("Append Dedup sync mode requires at least 1 primary key column"))
```